File: libs/language/src/lib/merge/processors/TypeAssigner.ts
import { BlueNode } from '../../model';
import { NodeProvider } from '../../NodeProvider';
import { MergingProcessor } from '../MergingProcessor';
import { NodeToMapListOrValue } from '../../utils/NodeToMapListOrValue';
import { NodeTypes } from '../../utils';

/**
 * Assigns types from source to target nodes with subtype validation
 */
export class TypeAssigner implements MergingProcessor {
  process(
    target: BlueNode,
    source: BlueNode,
    nodeProvider: NodeProvider,
  ): BlueNode {
    const targetType = target.getType();
    const sourceType = source.getType();
    let newTarget = target;

    if (targetType === undefined) {
      if (
        sourceType !== undefined &&
        source.getValue() !== undefined &&
        NodeTypes.isAnonymousCoreAlias(sourceType)
      ) {
        const sourceTypeStr = NodeToMapListOrValue.get(sourceType);
        throw new Error(
          `Anonymous core type alias '${JSON.stringify(
            sourceTypeStr,
          )}' is not interchangeable with its registry core type.`,
        );
      }
      newTarget = target.cloneShallow().setType(sourceType);
    } else if (sourceType !== undefined) {
      const isSubtypeResult = NodeTypes.isSubtype(
        sourceType,
        targetType,
        nodeProvider,
      );
      if (!isSubtypeResult) {
        const sourceTypeStr = NodeToMapListOrValue.get(sourceType);
        const targetTypeStr = NodeToMapListOrValue.get(targetType);
        throw new Error(
          `The source type '${JSON.stringify(
            sourceTypeStr,
          )}' is not a subtype of the target type '${JSON.stringify(
            targetTypeStr,
          )}'.`,
        );
      }
      newTarget = target.cloneShallow().setType(sourceType);
    }
    return newTarget;
  }
}

File: references/java-sdk/src/main/java/blue/language/merge/processor/TypeAssigner.java
package blue.language.merge.processor;

import blue.language.*;
import blue.language.merge.MergingProcessor;
import blue.language.merge.NodeResolver;
import blue.language.model.Node;
import blue.language.utils.NodeToMapListOrValue;

import static blue.language.utils.Types.isSubtype;

public class TypeAssigner implements MergingProcessor {

    @Override
    public void process(Node target, Node source, NodeProvider nodeProvider, NodeResolver nodeResolver) {
        Node targetType = target.getType();
        Node sourceType = source.getType();
        if (targetType == null)
            target.type(sourceType);
        else if (sourceType != null) {
            boolean isSubtype = isSubtype(sourceType, targetType, nodeProvider);
            if (!isSubtype) {
                String errorMessage = String.format("The source type '%s' is not a subtype of the target type '%s'.",
                        NodeToMapListOrValue.get(sourceType), NodeToMapListOrValue.get(targetType));
                throw new IllegalArgumentException(errorMessage);
            }
            target.type(sourceType);
        }
    }
}
